### PR TITLE
Add custom dup method for DateTimeIndex

### PR DESCRIPTION
**Pull request by Deepakkoli93**
https://github.com/v0dro/daru/issues/50 Added a test similar to `recode` for a dataframe which does not have a `DataTimeIndex`

### CODE REVIEW

File: spec/dataframe_spec.rb [hunk 1]
```diff
@@ -943,6 +947,35 @@
     end
   end
 
+  context "#recode_date_time" do
```
Comment (v0dro) [inline spec/dataframe_spec.rb]: The contents of the `context` specify the method that is being tested. There is no method called `#recode_date_time`.

You should create `it` blocks inside the `context "#recode"` and add the tests there. Also, when you do add this to the `context "#recode"` block, try to write the `before { ... }` without repeating dataframes.

PS: The string passed to `it` should describe exactly what the test does. Read this guide on rspec: http://betterspecs.org/

**Comment by Deepakkoli93** [comment]
I made the changes as you suggested. The new code looks like this
```ruby
  context "#recode" do
    before do
      @ans_vector = Daru::DataFrame.new({b: [21,22,23,24,25], a: [11,12,13,14,15],
        c: [21,32,43,54,65]}, order: [:a, :b, :c],
        index: [:one, :two, :three, :four, :five])

      @ans_rows = Daru::DataFrame.new({b: [121, 144, 169, 196, 225], a: [1,4,9,16,25],
        c: [121, 484, 1089, 1936, 3025]}, order: [:a, :b, :c],
        index: [:one, :two, :three, :four, :five])

      @data_frame_date_time = @data_frame.dup
      @data_frame_date_time.index = Daru::DateTimeIndex.date_range(start:"2016-02-11", periods:5)

      @ans_vector_date_time = Daru::DataFrame.new({b: [21,22,23,24,25], a: [11,12,13,14,15],
        c: [21,32,43,54,65]}, order: [:a, :b, :c],
        index: Daru::DateTimeIndex.date_range(start:"2016-02-11", periods:5))

      @ans_rows_date_time = Daru::DataFrame.new({b: [121, 144, 169, 196, 225], a: [1,4,9,16,25],
        c: [121, 484, 1089, 1936, 3025]}, order: [:a, :b, :c],
        index: Daru::DateTimeIndex.date_range(start:"2016-02-11", periods:5))
    end

    it "maps over the vectors of a DataFrame and returns a DataFrame" do
      ret = @data_frame.recode do |vector|
        vector.map! { |e| e += 10}
      end

      expect(ret).to eq(@ans_vector)
    end

    it "maps over the rows of a DataFrame and returns a DataFrame" do
      ret = @data_frame.recode(:row) do |row|
        expect(row.class).to eq(Daru::Vector)
        row.map! { |e| e*e }
      end

      expect(ret).to eq(@ans_rows)
    end

    it "maps over the vectors of a DataFrame with DateTimeIndex and returns a DataFrame with DateTimeIndex" do
      ret = @data_frame_date_time.recode do |vector|
        vector.map! { |e| e += 10}
      end

      expect(ret).to eq(@ans_vector_date_time)
    end

    it "maps over the rows of a DataFrame with DateTimeIndex and returns a DataFrame with DateTimeIndex" do
      ret = @data_frame_date_time.recode(:row) do |row|
        expect(row.class).to eq(Daru::Vector)
        row.map! { |e| e*e }
      end

      expect(ret).to eq(@ans_rows_date_time)
    end

  end
```

Just wanted to run it by you before pushing.

**Comment by v0dro** [comment]
Looks good.

**Comment by Deepakkoli93** [comment]
I've pushed the new changes.

File: lib/daru/date_time/index.rb [hunk 2]
```diff
@@ -268,6 +268,16 @@ def initialize *args
...
     end
 
+    # dup method for DateTimeIndex
+    
+    def dup
+      data = []
+      @data.each do |i|
+        data << i[0]
+      end
+      Daru::DateTimeIndex.new(data, :freq => @offset)
```
Comment (v0dro) [inline lib/daru/date_time/index.rb]: You can just write this as `Daru::DateTimeIndex.new(@data.transpose[0], :freq => @offset)`

File: lib/daru/date_time/index.rb [hunk 3]
```diff
@@ -268,6 +268,16 @@ def initialize *args
       @periods   = data.size
     end
 
+    # dup method for DateTimeIndex
+
```
Comment (v0dro) [inline lib/daru/date_time/index.rb]: there should be no newline between method and docs.

**Comment by Deepakkoli93** [comment]
Made the changes as you suggested

**Pull request merged**